src/compiler/types/tests/generate-app-types.spec.ts

```diff
@@ -0,0 +1,153 @@
+      srcDir: '/',
+    });
+    compilerCtx = {
+      ...mockCompilerCtx(config),
+      fs: mockFs as any,
+    };
+    buildCtx = mockBuildCtx(config, compilerCtx);
+
+    mockFs.writeFile.mockResolvedValueOnce({ changedContent: true });
+  });
+
+  afterEach(() => {
+    jest.resetAllMocks();
+  });
+
+  it('should generate a type declaration file without custom event types', async () => {
+    const compilerComponentMeta = stubComponentCompilerMeta({
+      tagName: 'my-component',
+      componentClassName: 'MyComponent',
+      events: [],
+    });
+    buildCtx.components = [compilerComponentMeta];
+
+    await generateAppTypes(config, compilerCtx, buildCtx, 'src');
+
+    expect(mockFs.writeFile).toHaveBeenCalledWith(
+      '/components.d.ts',
+      `/* eslint-disable */
+/* tslint:disable */
+/**
+ * This is an autogenerated file created by the Stencil compiler.
+ * It contains typing information for all components that exist in this project.
+ */
+import { HTMLStencilElement, JSXBase } from "@stencil/core/internal";
+export namespace Components {
+        interface MyComponent {
+        }
+}
+declare global {
+        interface HTMLMyComponentElement extends Components.MyComponent, HTMLStencilElement {
+        }
+        var HTMLMyComponentElement: {
+                prototype: HTMLMyComponentElement;
+                new (): HTMLMyComponentElement;
+        };
+        interface HTMLElementTagNameMap {
+                "my-component": HTMLMyComponentElement;
+        }
+}
+declare namespace LocalJSX {
+      interface MyComponent {
+        }
+        interface IntrinsicElements {
+              "my-component": MyComponent;
+        }
+}
+export { LocalJSX as JSX };
+declare module "@stencil/core" {
+        export namespace JSX {
+                interface IntrinsicElements {
+                        "my-component": LocalJSX.MyComponent & JSXBase.HTMLAttributes<HTMLMyComponentElement>;
+                }
+        }
+}
+`,
+      {
+        immediateWrite: true,
+      }
+    );
+  });
+
+  it('should generate a type declaration file with custom event types', async () => {
+    const compilerComponentMeta = stubComponentCompilerMeta({
+      tagName: 'my-component',
+      componentClassName: 'MyComponent',
+      hasEvent: true,
+      events: [stubComponentCompilerEvent()],
+    });
+    buildCtx.components = [compilerComponentMeta];
+
+    await generateAppTypes(config, compilerCtx, buildCtx, 'src');
+
+    expect(mockFs.writeFile).toHaveBeenCalledWith(
+      '/components.d.ts',
+      `/* eslint-disable */
+/* tslint:disable */
+/**
+ * This is an autogenerated file created by the Stencil compiler.
+ * It contains typing information for all components that exist in this project.
+ */
+import { HTMLStencilElement, JSXBase } from "@stencil/core/internal";
+import { UserImplementedEventType } from "./some/stubbed/path/resources";
+export { UserImplementedEventType } from "./some/stubbed/path/resources";
```

Comment:
Can we add a few additional test scenarios here?

1. Exporting multiple types from a single location to exercise the `join()` on multiple types
2. A type that is defined/used in multiple components (ideally with a different hierarchy). Consider the example below where `common-types` has a type that is used in both `level-1-component` and `level-2-component`. This should exercise path resolution of types we export
```
src
├── components
│   └── level-1
│       ├── level-1-component.tsx
│       └── level-2
│            └── level-2-component.tsx
└── utils
    └── common-types.ts
```

3. Naming collisions - If we have two components that each define and use a custom event type of the same name (but a different type), what gets exported? e.g.
```ts
// src/component-a.tsx

export type MyEvent = string;
```
```ts
// src/component-b.tsx

export type MyEvent = number;
```

Reply:
Managed to get the name collisions for local imports resolved. Added those test cases and a few more for prop types based on the comments in [this thread](https://github.com/ionic-team/stencil/pull/3612#discussion_r975424761)